### PR TITLE
Ordering is Not Guaranteed with a Simple Kafka Consumer

### LINKED ISSUE #228: Ordering is Not Guaranteed with a Simple Kafka Consumer
## Is this a bug? Please provide steps to reproduce, a failing test etc.

While iterating over the messages of a partition consumer, we spin up some work into another goroutine: https://github.com/beatlabs/patron/blob/7c8e35c9547d99025d53a97b25526496849c8068/component/async/kafka/simple/simple.go#L132
There is no guarantee that a goroutine created before another will be triggered and complete before. Hence, we cannot guarantee the delivery of the messages to the Patron clients. 

For information, I was able to trigger this issue while testing `TestSimpleConsume` for example.

## Describe the solution

The solution would be to remove the extra goroutine, just like what was done in the group consumer:
https://github.com/beatlabs/patron/blob/726c0299dcf9f99b14357429e54b7eca89317c7c/component/async/kafka/group/group.go#L171:L177

### PR DESCRIPTION
**Pull request by teivah**
## Which problem is this PR solving?

Guaranteed ordering with a simple Kafka consumer: Closes #228.

## Short description of the changes

Removing the extra goroutine created while iterating over the partition consumer messages.

### CODE REVIEW

**Review by mantzas — DISMISSED** [review]

**Review by drakos74 — DISMISSED** [review]

**Comment by drakos74** [comment]
looks ok,
just to be aware, now in the case of an error, the consumer will stop processing
(because we return from the Consumer routine).
Maybe worth to investigate the behaviour of this a bit more, and if needed add to the doc,
or create a follow-up issue, to enable also the previous behaviour.
WDYT ? @teivah @mantzas  ?

File: component/async/kafka/simple/simple.go [hunk 1]
```diff
@@ -118,16 +118,14 @@ func (c *consumer) Consume(ctx context.Context) (<-chan async.Message, <-chan er
...
-							return
-						}
-						kafka.MessageStatusCountInc(kafka.MessageDecoded, "", message.Topic)
-						chMsg <- msg
-					}(m)
+					msg, err := kafka.ClaimMessage(ctx, m, c.config.DecoderFunc, nil)
+					if err != nil {
+						kafka.MessageStatusCountInc(kafka.MessageClaimErrors, "", m.Topic)
+						chErr <- err
+						return
```
Comment (drakos74) [inline component/async/kafka/simple/simple.go]: the more i think about it ...
it seems we could avoid return at this point

**Comment by drakos74** [comment]
> looks ok,
> just to be aware, now in the case of an error, the consumer will stop processing
> (because we return from the Consumer routine).
> Maybe worth to investigate the behaviour of this a bit more, and if needed add to the doc,
> or create a follow-up issue, to enable also the previous behaviour.
> WDYT ? @teivah @mantzas ?

@teivah , would you mind creating a test to assert these assumptions ? And add it to the PR ?
So the behaviour becomes clear ?

File: component/async/kafka/simple/simple.go (hunk 1 above)
Comment (teivah) [reply component/async/kafka/simple/simple.go]: Shall we? I'm puzzled... On one hand, I understand that it changes the current behaviour. On the other hand, the group consumer already stops the message consumption when it faced a `ClaimMessage` error: https://github.com/beatlabs/patron/blob/726c0299dcf9f99b14357429e54b7eca89317c7c/component/async/kafka/group/group.go#L174

Comment (drakos74) [reply component/async/kafka/simple/simple.go]: yeah , it s not clear to me either what exactly this change will effect (probably also nothing).
Do you think we have some test that demonstrates this ?
Or should we add one ?

**Review by drakos74 — APPROVED** [review]

**Review by mantzas — APPROVED** [review]

**Pull request merged**